notebook/notebook/notebook.py

```diff
@@ -131,10 +171,153 @@ def external_url_for(path):
     return url + path
 
 
+def forbidden():
+    return 'Forbidden', 403
+
+
 @app.route('/healthcheck')
 def healthcheck():
     return '', 200
 
+##### New notebook methods meant for consumption from agnostic client#####
+###### Methods that do not call the auth-gateway /verify method must not be directly public #######
+@app.route('/verify/<svc_name>/', methods=['GET'])
+def validate_get(svc_name):
+    access_token = request.args.get('authorization')
+    token = request.args.get('token')
+
+    if not access_token:
+        return forbidden()
+
+    resp = requests.get(f'http://{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
```

Comment:
If I understand correctly, all services are post-authentication, ergo, we can trust some header or query parameter set by the gateway nginx.

Reply:
Pursuant to this, I think the first step towards the authentication-based notebook is to decouple gateway and nginx by placing a second nginx in a second container in the same pod as the notebook service and directing the `notebook` k8s `service` to the nginx container rather than to the notebook container.

I started down this path, but you should take it to completion. My branch in this direction is kind of a mess of other entangled, failed ideas and merge conflicts, but [you can peek if you want](https://github.com/hail-is/hail/compare/master...danking:notebook-make-run#diff-15d2783b37ed27bbfd1edfff81b39b92).

You'll need to revert `gateway`'s `notebook` `server` block to look like all the other services. The `notebook` `server` block will need to move into this directory. I'd add a subdirectory `notebook/nginx` which contains a Dockerfile and the nginx configuration file. You'll need to modify the `Makefile` to build this image as well. `deployment.yaml.in` will need to change:
```
kind: Deployment
...
      containers:
      - name: notebook
        image: @notebook-leader-image@
        ports:
        - containerPort: 5000
        volumeMounts:
          - mountPath: /notebook-secrets
            readOnly: true
            name: notebook-secrets
      - name: notebook
        image: @notebook-nginx-image@
        ports:
        - containerPort: 80
---
kind: Service
...
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
```

The nginx configuration should redirect to `http://localhost:5000` instead of `http://notebook` when redirecting to the notebook leader.

Reply:
To first comment: yes, except in the case that only 1 auth_request is possible, e.g when we have only 1 nginx interaction before redirect. The decoupling will happen - thanks for the tips and partial code - after the workshop most likely.